src/linear_algebra/affine_space/restrict.lean

```diff
@@ -0,0 +1,93 @@
+/-
+Copyright (c) 2022 Paul Reichert. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul Reichert
+-/
+import linear_algebra.affine_space.affine_subspace
+
+/-!
+# Affine map restrictions
+
+This file defines restrictions of affine maps.
+
+## Main definitions
+
+* The domain and codomain of an affine map can be restricted using
+  `affine_map.restrict`.
+
+## Main theorems
+
+* The associated linear map of the restriction is the restriction of the
+  linear map associated to the original affine map.
+* The restriction is injective if the original map is injective.
+* The restriction in surjective if the codomain is the image of the domain.
+-/
+
+variables {k V₁ P₁ V₂ P₂ : Type} [ring k]
+  [add_comm_group V₁] [add_comm_group V₂]
+  [module k V₁] [module k V₂]
+  [add_torsor V₁ P₁] [add_torsor V₂ P₂]
+
+include V₁ V₂
+
+/- not an instance because it loops with `nonempty` -/
+lemma affine_subspace.nonempty_map {E : affine_subspace k P₁} [Ene : nonempty E]
+  {φ : P₁ →ᵃ[k] P₂} : nonempty (E.map φ) :=
+begin
+  obtain ⟨x, hx⟩ := id Ene,
+  refine ⟨⟨φ x, affine_subspace.mem_map.mpr ⟨x, hx, rfl⟩⟩⟩,
+end
+
+local attribute [instance, nolint fails_quickly] affine_subspace.nonempty_map
+local attribute [instance, nolint fails_quickly] affine_subspace.to_add_torsor
+
+/-- Restrict domain and codomain of an affine map to the given submodules. -/
+def affine_map.restrict
+  (φ : P₁ →ᵃ[k] P₂) {E : affine_subspace k P₁} {F : affine_subspace k P₂}
+  [nonempty E] [nonempty F]
+  (hEF : E.map φ ≤ F) : E →ᵃ[k] F :=
+begin
+  refine ⟨_, _, _⟩,
+  { exact λ x, ⟨φ x, hEF $ affine_subspace.mem_map.mpr ⟨x, x.property, rfl⟩⟩ },
+  { refine φ.linear.restrict' _,
+    rw [←affine_subspace.map_direction],
+    exact affine_subspace.direction_le hEF },
+  { intros p v,
+    simp only [subtype.ext_iff, subtype.coe_mk, affine_subspace.coe_vadd],
+    apply affine_map.map_vadd },
+end
+
+lemma affine_map.restrict.coe_apply
+  (φ : P₁ →ᵃ[k] P₂) {E : affine_subspace k P₁} {F : affine_subspace k P₂}
+  [nonempty E] [nonempty F]
+  (hEF : E.map φ ≤ F) (x : E) :
+  ↑(φ.restrict hEF x) = φ x := rfl
+
+lemma affine_map.restrict.linear
+  (φ : P₁ →ᵃ[k] P₂) {E : affine_subspace k P₁} {F : affine_subspace k P₂}
+  [nonempty E] [nonempty F]
+  (hEF : E.map φ ≤ F) :
+  (φ.restrict hEF).linear = φ.linear.restrict'
+    (by { rw [←affine_subspace.map_direction], exact affine_subspace.direction_le hEF }) := rfl
```

Comment:
Does `simps` generate these? Or you prefer spelling them out explicitly instead?

Reply:
Actually, no. It would be generated if I added `@[simps]` to `affine_map.restrict`, but I believe this is not desirable because the unfolded definition looks harder to manipulate. The function `linear_map.restrict` does not have `@[simps]` either -- the only reason why `linear_map.restrict'` has been afforded the attribute is that it's just a shorthand for `linear_map.restrict`.